### PR TITLE
prefer std::variant to boost::variant

### PR DESCRIPTION
**Pull request by yaozongyou**
Signed-off-by: Yao Zongyou <yaozongyou@vip.qq.com>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,13 +4,9 @@ project(dmclock CXX)
...
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake/modules")
 
-if (NOT(BOOST_FOUND))
-  find_package(Boost REQUIRED)
-endif()
-
 find_package(Threads)
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 17)
```
Comment (tchaikov) [inline CMakeLists.txt]: this setting is introduced in cmake 3.8, but we need to support cmake 3.5.1, please refer to  

https://github.com/ceph/ceph/blob/5a3344f0e52c05c5f14c326e676bb9e5a065b82b/src/CMakeLists.txt#L102-L125

on how to appease cmake 3.5.1

**Comment by cbodley** [comment]
are there dmclock users outside of ceph that care about c++11 support?

**Comment by ivancich** [comment]
I have a similar question as @cbodley and I'm inclined to think maintaining c++11 support is worthwhile. ceph uses boost heavily; over 500 source files below src/ refer to "boost".

Or put another way, what specifically do we gain by this change?

**Comment by yaozongyou** [comment]
@ivancich the reason for creating this pr is in my environment, dmclock fail to build, i am using gcc7.2 cmake version 3.9.5 Boost version: 1.52.0.  Here is some part of the error message:
```bash
[root@localhost /data/home/richardyao/workspace/github/dmclock/build]# make 
[ 30%] Built target dmclock
[ 40%] Building CXX object test/CMakeFiles/dmclock-tests.dir/test_dmclock_server.cc.o
In file included from /opt/boost_1_52_0/boost/variant/variant.hpp:31:0,
                 from /opt/boost_1_52_0/boost/variant.hpp:17,
                 from /data/home/richardyao/workspace/github/dmclock/src/dmclock_server.h:42,
                 from /data/home/richardyao/workspace/github/dmclock/test/test_dmclock_server.cc:23:
/opt/boost_1_52_0/boost/variant/detail/initializer.hpp: In instantiation of ‘static int boost::detail::variant::make_initializer_node::apply<BaseIndexPair, Iterator>::initializer_node::initialize(void*, boost::detail::variant::make_initializer_node::apply<BaseIndexPair, Iterator>::initializer_node::param_T) [with BaseIndexPair = boost::mpl::pair<boost::detail::variant::initializer_root, mpl_::int_<0> >; Iterator = boost::mpl::l_iter<boost::mpl::list2<crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn, double> >; boost::detail::variant::make_initializer_node::apply<BaseIndexPair, Iterator>::initializer_node::param_T = const crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn&]’:
/opt/boost_1_52_0/boost/variant/variant.hpp:1328:38:   required from ‘void boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>::convert_construct(T&, int, mpl_::false_) [with T = const crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T0_ = crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T1 = double; T2 = boost::detail::variant::void_; T3 = boost::detail::variant::void_; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; mpl_::false_ = mpl_::bool_<false>]’
/opt/boost_1_52_0/boost/variant/variant.hpp:1400:26:   required from ‘boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>::variant(const T&) [with T = crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T0_ = crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T1 = double; T2 = boost::detail::variant::void_; T3 = boost::detail::variant::void_; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_]’
/opt/boost_1_52_0/boost/variant/variant.hpp:1643:21:   required from ‘void boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>::assign(const T&) [with T = crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T0_ = crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T1 = double; T2 = boost::detail::variant::void_; T3 = boost::detail::variant::void_; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_]’
/opt/boost_1_52_0/boost/variant/variant.hpp:1653:15:   required from ‘boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>& boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>::operator=(const T&) [with T = crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T0_ = crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn; T1 = double; T2 = boost::detail::variant::void_; T3 = boost::detail::variant::void_; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_]’
/data/home/richardyao/workspace/github/dmclock/src/dmclock_server.h:1429:23:   required from ‘crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::pull_request(crimson::dmclock::Time)::<lambda(crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::PullReq&, crimson::dmclock::PhaseType)>::<lambda(const C&, crimson::dmclock::Cost, typename crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::super::RequestRef&)> [with C = int; R = Request; bool IsDelayed = true; bool U1 = false; unsigned int B = 2; crimson::dmclock::Cost = unsigned int; typename crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::super::RequestRef = std::unique_ptr<Request, std::default_delete<Request> >]’
/data/home/richardyao/workspace/github/dmclock/src/dmclock_server.h:1426:26:   required from ‘struct crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::pull_request(crimson::dmclock::Time)::<lambda(crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::PullReq&, crimson::dmclock::PhaseType)> [with C = int; R = Request; bool IsDelayed = true; bool U1 = false; unsigned int B = 2; typename crimson::dmclock::PriorityQueueBase<C, R, IsDelayed, U1, B>::RequestRef = std::unique_ptr<Request, std::default_delete<Request> >]::<lambda(const int&, using Cost = uint32_t, using RequestRef = class std::unique_ptr<Request, std::default_delete<Request> >&)>’
/data/home/richardyao/workspace/github/dmclock/src/dmclock_server.h:1433:4:   required from ‘crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::pull_request(crimson::dmclock::Time)::<lambda(crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::PullReq&, crimson::dmclock::PhaseType)> [with C = int; R = Request; bool IsDelayed = true; bool U1 = false; unsigned int B = 2; typename crimson::dmclock::PriorityQueueBase<C, R, IsDelayed, U1, B>::RequestRef = std::unique_ptr<Request, std::default_delete<Request> >]’
/data/home/richardyao/workspace/github/dmclock/src/dmclock_server.h:1422:6:   required from ‘struct crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::pull_request(crimson::dmclock::Time) [with C = int; R = Request; bool IsDelayed = true; bool U1 = false; unsigned int B = 2; crimson::dmclock::Time = double]::<lambda(struct crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq&, enum class crimson::dmclock::PhaseType)>’
/data/home/richardyao/workspace/github/dmclock/src/dmclock_server.h:1421:7:   required from ‘crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::PullReq crimson::dmclock::PullPriorityQueue<C, R, IsDelayed, U1, B>::pull_request(crimson::dmclock::Time) [with C = int; R = Request; bool IsDelayed = true; bool U1 = false; unsigned int B = 2; crimson::dmclock::Time = double]’
/data/home/richardyao/workspace/github/dmclock/test/test_dmclock_server.cc:208:33:   required from here
/opt/boost_1_52_0/boost/variant/detail/initializer.hpp:95:17: error: use of deleted function ‘crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn::Retn(const crimson::dmclock::PullPriorityQueue<int, Request, true>::PullReq::Retn&)’
                 new(dest) internal_T(operand);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

maybe i need to update my boost, thanks~

**Comment by cbodley** [comment]
> maybe i need to update my boost, thanks~

most likely, yeah - let us know if it still fails with a more recent boost

**Comment by ivancich** [comment]
@chardan That's very helpful. My sense is that maintaining c++11 compatibility is worth it for the time being but that moving to std::variant will be done in the future. (I don't know how much work would be involved with conditionally compiling one or the other; I'm focused on other things.)

**Comment by ivancich** [comment]
@yaozongyou Thank you for taking the initiative and working through the boost issues. I will close this pr as you advised.

**Pull request closed without merging**